For Node.js 10.x and 12.x the import.meta object is incompletely defined

- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [ ] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
 - Authors: @tdreyno @kamal @leemhenson @jonmagic @farcaller 

I have already asked the question part of this on Stack Overflow:  https://stackoverflow.com/questions/55998413/equivalent-for-dirname-in-typescript-when-using-target-module-esnext

The issue is that using TypeScript with @types/node etc installed, and I need to calculate the value for `__dirname`, the reference to `import.meta.url` cannot be compiled by the TypeScript compiler.

In the Node.js 12.x documentation it says to calculate `__dirname` from `import.meta.url`.

But referencing that variable in some TypeScript code causes the compilation to fail

```
tdn.ts:7:25 - error TS2339: Property 'url' does not exist on type 'ImportMeta'.

7 console.log(import.meta.url);
```

I found in `node_modules//typescript/lib/lib.es5.d.ts` this code:

```
/**
 * The type of `import.meta`.
 *
 * If you need to declare that a given property exists on `import.meta`,
 * this type may be augmented via interface merging.
 */
interface ImportMeta {
}
```

So.... 

Clearly the `ImportMeta` interface is not properly defined.  For Node.js there is a `url` field that is clearly defined in the Node.js documentation.

The definition in the TypeScript source probably should not change.  But the definition in `@types/node` should probably change to match reality.

--- a/types/node/globals.d.ts
+++ b/types/node/globals.d.ts
@@ -154,6 +154,10 @@ interface String {
     trimRight(): string;
 }
 
+interface ImportMeta {
+    url: string;
+}
+
 /*-----------------------------------------------*
  *                                               *
  *                   GLOBAL                      *

--- a/types/node/node-tests.ts
+++ b/types/node/node-tests.ts
@@ -562,65 +562,6 @@ import Module = require("module");
     const end2: string = decoder1.end(new Buffer('test'));
 }
 
-////////////////////////////////////////////////////
-/// vm tests : https://nodejs.org/api/vm.html
-////////////////////////////////////////////////////
-
-{
-    {
-        const sandbox = {
-            animal: 'cat',
-            count: 2
-        };
-
-        const context = vm.createContext(sandbox);
-        console.log(vm.isContext(context));
-        const script = new vm.Script('count += 1; name = "kitty"');
-
-        for (let i = 0; i < 10; ++i) {
-            script.runInContext(context);
-        }
-
-        console.log(util.inspect(sandbox));
-
-        vm.runInNewContext('count += 1; name = "kitty"', sandbox);
-        console.log(util.inspect(sandbox));
-    }
-
-    {
-        const sandboxes = [{}, {}, {}];
-
-        const script = new vm.Script('globalVar = "set"');
-
-        sandboxes.forEach((sandbox) => {
-            script.runInNewContext(sandbox);
-            script.runInThisContext();
-        });
-
-        console.log(util.inspect(sandboxes));
-
-        const localVar = 'initial value';
-        vm.runInThisContext('localVar = "vm";');
-
-        console.log(localVar);
-    }
-
-    {
-        vm.runInThisContext('console.log("hello world"', './my-file.js');
-    }
-
-    {
-        const fn: Function = vm.compileFunction('console.log("test")', [], {
-            parsingContext: vm.createContext(),
-            contextExtensions: [{
-                a: 1,
-            }],
-            produceCachedData: false,
-            cachedData: Buffer.from('nope'),
-        });
-    }
-}
-
 /////////////////////////////////////////////////////
 /// Timers tests : https://nodejs.org/api/timers.html
 /////////////////////////////////////////////////////

--- a/types/node/test/vm.ts
+++ b/types/node/test/vm.ts
@@ -1,0 +1,62 @@
+import { createContext, isContext, Script, runInNewContext, runInThisContext, compileFunction } from 'vm';
+import { inspect } from 'util';
+
+{
+    const sandbox = {
+        animal: 'cat',
+        count: 2
+    };
+
+    const context = createContext(sandbox, {
+        name: 'test',
+        origin: 'file://test.js',
+        codeGeneration: {
+            strings: true,
+            wasm: true,
+        },
+    });
+    console.log(isContext(context));
+    const script = new Script('count += 1; name = "kitty"');
+
+    for (let i = 0; i < 10; ++i) {
+        script.runInContext(context);
+    }
+
+    console.log(inspect(sandbox));
+
+    runInNewContext('count += 1; name = "kitty"', sandbox);
+    console.log(inspect(sandbox));
+}
+
+{
+    const sandboxes = [{}, {}, {}];
+
+    const script = new Script('globalVar = "set"');
+
+    sandboxes.forEach((sandbox) => {
+        script.runInNewContext(sandbox);
+        script.runInThisContext();
+    });
+
+    console.log(inspect(sandboxes));
+
+    const localVar = 'initial value';
+    runInThisContext('localVar = "vm";');
+
+    console.log(localVar);
+}
+
+{
+    runInThisContext('console.log("hello world"', './my-file.js');
+}
+
+{
+    const fn: Function = compileFunction('console.log("test")', [], {
+        parsingContext: createContext(),
+        contextExtensions: [{
+            a: 1,
+        }],
+        produceCachedData: false,
+        cachedData: Buffer.from('nope'),
+    });
+}

--- a/types/node/tsconfig.json
+++ b/types/node/tsconfig.json
@@ -67,6 +67,7 @@
         "test/tty.ts",
         "test/util.ts",
         "test/v8.ts",
+        "test/vm.ts",
         "test/worker_threads.ts",
         "test/zlib.ts"
     ],

--- a/types/node/vm.d.ts
+++ b/types/node/vm.d.ts
@@ -49,13 +49,43 @@ declare module "vm" {
          */
         contextExtensions?: Object[];
     }
+
+    interface CreateContextOptions {
+        /**
+         * Human-readable name of the newly created context.
+         * @default 'VM Context i' Where i is an ascending numerical index of the created context.
+         */
+        name?: string;
+        /**
+         * Corresponds to the newly created context for display purposes.
+         * The origin should be formatted like a `URL`, but with only the scheme, host, and port (if necessary),
+         * like the value of the `url.origin` property of a URL object.
+         * Most notably, this string should omit the trailing slash, as that denotes a path.
+         * @default ''
+         */
+        origin?: string;
+        codeGeneration?: {
+            /**
+             * If set to false any calls to eval or function constructors (Function, GeneratorFunction, etc)
+             * will throw an EvalError.
+             * @default true
+             */
+            strings?: boolean;
+            /**
+             * If set to false any attempt to compile a WebAssembly module will throw a WebAssembly.CompileError.
+             * @default true
+             */
+            wasm?: boolean;
+        };
+    }
+
     class Script {
         constructor(code: string, options?: ScriptOptions);
         runInContext(contextifiedSandbox: Context, options?: RunningScriptOptions): any;
         runInNewContext(sandbox?: Context, options?: RunningScriptOptions): any;
         runInThisContext(options?: RunningScriptOptions): any;
     }
-    function createContext(sandbox?: Context): Context;
+    function createContext(sandbox?: Context, options?: CreateContextOptions): Context;
     function isContext(sandbox: Context): boolean;
     function runInContext(code: string, contextifiedSandbox: Context, options?: RunningScriptOptions | string): any;
     function runInNewContext(code: string, sandbox?: Context, options?: RunningScriptOptions | string): any;

